File: crates/klyron_engine/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::EngineBuilder::Impl {
public:
  Klyron::EngineConfig config;
  bool verbose = false;
};

Klyron::EngineBuilder::Klyron::EngineBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::EngineBuilder& Klyron::EngineBuilder::withConfig(const Klyron::EngineConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::EngineBuilder& Klyron::EngineBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::EngineBuilder::Klyron::EngineInstance Klyron::EngineBuilder::build() {
  Klyron::EngineInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_engine/bindings/rust/builder.rs
//! Builder pattern for klyron_engine

use crate::types::Klyron::EngineConfig;

#[derive(Debug, Default)]
pub struct Klyron::EngineBuilder {
    config: Option<Klyron::EngineConfig>,
    verbose: bool,
}

impl Klyron::EngineBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::EngineConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::EngineInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::EngineInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::EngineInstance {
    pub config: Klyron::EngineConfig,
    pub verbose: bool,
}
